### DISCUSSION TITLE
Need to replicate SectionByShape behavior

### DISCUSSION THREAD

**Discussion by ovaisa** [iTwin/itwinjs-core#5388] [General] [answered]
I want to implement a workflow similar to ```CoreTools.sectionByShapeCommandItemDef```.  The workflows is:

1. Let user select a shape
2. Clip according to the selected shape
3. Display Clip decorator so that user can interactively modify clipped region

I wrote (prototype) code below to illustrate my requirement

By clicking ```Clip Shape``` button, I can do steps 1 and 2 (by running ```ViewClipByShapeTool```, pls see ```onClipShape()```)
and after selecting shape, I can do step 3 by clicking ```Show Decorator``` (pls see ```addDecorations()```, copied from Bentley samples).  By clicking these two buttons separately I get the workflow I want.

My requirement is to do these three steps with a single click.  I tried calling ```addDecorations()``` in ```onClipShape()``` but that seems to cancel the ```ViewClipByShapeTool``` and none of the steps can be performed.

Can you please guide how I can do these three steps with a single click.
```
const AppUiElements: React.FunctionComponent = () => {

  const addDecorators = () => {   // from VolumeQuery sample
    const decorationProvider: ViewClipDecorationProvider = ViewClipDecorationProvider.create();
    decorationProvider.clearDecorationOnDeselect = false;
    decorationProvider.showDecoration(IModelApp.viewManager.selectedView!);
    void IModelApp.toolAdmin.startDefaultTool();
  }
    
  const onClipShape = async () => {
    await IModelApp.tools.run(ViewClipByShapeTool.toolId);
//    addDecorators();   // calling decorator seems to cancel ViewClipByShapeTool
  }

  const onShowDecorator = async () => {
    addDecorators();
  }

  return (
    <div>
      <Button onClick={onClipShape}>Clip Shape</Button>
      <Button onClick={onShowDecorator}>Show Decorator</Button>
    </div>
  )
}
```

**Comment by bbastings** [comment]
If you supply a ViewClipDecorationProvider to the ViewClipByShapeTool, the default behavior is to create and select the view clip decoration.

await IModelApp.tools.run(ViewClipByShapeTool.toolId, **ViewClipDecorationProvider.create()**);

If you are writing your own sub-class of ViewClipTool, you just need to follow the same pattern.

**Reply by ovaisa** [reply]
Thank you Brian, this works exactly as I wanted!!!

I have a second workflow which is very similar, except in this workflow I don't want to clip the model (steps 1 and 3 only, no step 2)

We want to let the user select a shape and then show clip decorator to let user modify the shape.  The goal is to identify elements within the user defined volume and then do certain calcs on them (e.g. find total weight of elements within the volume).  In this case we don't want to clip the model, just show the decorator 'volume' in the model. This requirement is same as VolumeQuery sample with ```Clip Volume``` toggle off.

I set ```clipVolume``` ```viewFlag``` to ```false``` before calling ```IModelApp.tools.run()``` in the above code (same as VolumeQuery sample) but the model is still clipped.   My modified code (```onClipShape()```) is below.  Would appreciate if you can guide on how I can show the decorator without clipping the model.
```
const onClipShape = async () => {
  const vp = IModelApp.viewManager.selectedView!;
  vp.view.viewFlags = vp.view.viewFlags.with("clipVolume", false);
  await IModelApp.tools.run(ViewClipByShapeTool.toolId, ViewClipDecorationProvider.create());
}
```

PS: I am not writing my own sub-class of ViewClipTool, my actual code is pretty much same as the prototype code shown here.

**Reply by bbastings** [reply]
Turning off the view flag before running the tool doesn't work because ViewClipByShapeTool will turn it on when you accept the shape (see below).

```
  public override async onDataButtonDown(ev: BeButtonEvent): Promise<EventHandled> {
    if (undefined === this.targetView)
      return EventHandled.No;

    if (this._points.length > 1 && !ev.isControlKey) {
      const points = this.getClipPoints(ev);
      if (points.length < 3)
        return EventHandled.No;

      const transform = Transform.createOriginAndMatrix(points[0], this._matrix);
      transform.multiplyInversePoint3dArrayInPlace(points);
      ViewClipTool.enableClipVolume(this.targetView); // <- HERE
      if (!ViewClipTool.doClipToShape(this.targetView, points, transform, this._zLow, this._zHigh))
        return EventHandled.No;
      if (undefined !== this._clipEventHandler)
        this._clipEventHandler.onNewClip(this.targetView);
      await this.onReinitialize();
      return EventHandled.Yes;
    }
```

You are trying to use this tool for a purpose for which it was not intended.

**Reply by ovaisa** [reply]
Regarding your question on use of this tool, it could be that this is not the right tool for this workflow.  The workflow I want is same as the QueryVolume sample and I am trying to use the same tools in that sample (which also uses Clip tools even though there is no clipping when Clip Volume toggle 'off').

If there is another tool that can achieve this workflow, I can use that (and is perhaps better as it seems this is not the right tool).  My requirement is basically:

1.  Let user define a volume by selecting a shape and a height - same a clip decorator defines a volume/region - but no clipping
2. Query and find elements in the volume - same as VolumeQuery sample
3. Do the calculation on the elements returned in the query

If there is a different tool that I should be using, please suggest, I will be happy to try.

**Reply by bbastings** [reply, marked as answer]
While I can understand wanting to use the existing tool and clip decoration since it's so close to what you want, there will obviously be times where creating your own tool to fully customize the behavior is desirable, this is one of those times.

You do have the option of sub-classing from the existing tool if you are happy with just turning the view flag off. Something like the following will likely suffice (as far as just creating/showing the clip decoration).

```
class MyTestClipByShapeTool extends ViewClipByShapeTool {
  public static override toolId = "MyTestClipByShape";

  public static disableClipVolume(viewport: Viewport): boolean {
    if (!viewport.viewFlags.clipVolume)
      return false;

    viewport.viewFlags = viewport.viewFlags.with("clipVolume", false);
    return true;
  }

  public override async onDataButtonDown(ev: BeButtonEvent): Promise<EventHandled> {
    if (EventHandled.No === await super.onDataButtonDown(ev))
      return EventHandled.No;
    if (ev.viewport)
      MyTestClipByShapeTool.disableClipVolume(ev.viewport); // NOTE: Tool has exited, don't call instance methods!
    return EventHandled.Yes;
  }
}
```

**Reply by ovaisa** [reply]
Thank you Brian, seems fairly straight forward, I will certainly try to implement this.  Thank you for your suggestion.

Though I expect that this sub-tool has to be registered and probably requires additional management so that it can be handled by IModelApp.tools.run().   Is there an sample app that has a similar custom tool implementation or any documentation that gives an overview of the tool development including registry etc.

Would appreciate if you can point me to such material.

Thanks,

Ovais

**Reply by ovaisa** [reply]
Thank you Brian, this works exactly I wanted!!!  Much appreciate your the code you provided and you support in getting this workflow to work.

Ovais
